### ISSUE TITLE
Millions of Workflow Instances

### ISSUE THREAD

**Issue by generatives** [elsa-workflows/elsa-core#306]
Elsa seems like a very interesting and powerful tool, I would love to put it to use. My only concern right now is how well it behaves at scale. Based on the example given Elsa seems to be built for smaller scale internal tools, I am looking for something to process user actions. Could Elsa handle millions of workflow instances?

**Comment by sfmskywalker** [comment]
Honestly, I have no idea :D I haven't done any serious load testing and I don;t know if anyone is using it at scale.

That being said, I don't see any inherent limitation that would prevent Elsa from operating at scale. Depending on what your workflows do when loaded in memory and how long they stay in memory, you should be able to handle an unlimited number of workflow instances. When executing a workflow, it either completes or gets suspended when it reaches a blocking activity. Which means the workflow instance will get persisted.

Loading workflow instances should be fine too, assuming the database has proper indices for the fields by which workflow instances are queried. Which (in theory, I still make mistakes) is the case with the DB providers Elsa ships with.

In any case, thanks for raising the question. I will keep this issue open as a way to remind myself I should setup load testing.

**Comment by spokeDev** [comment]
I noticed this while testing yesterday. 

On each sweep of the TimersHostedService, workflows with a blocking activity will have their ActivityInstanceEntities removed and re-added. This means that the integer field, ActivityInstanceEntity.Id, auto increments with every check on the blocking activity. See EntityFrameworkCoreWorkflowInstanceStore.SaveAsync().
So you get 2,147,483,647 before you run into an issue. 

Not an issue for our implementation, but if you have millions of workflows with blocking activities on a tight sweep interval you might run into a problem down the road.

**Comment by sfmskywalker** [comment]
Thanks for raising this. I can see how this becomes an issue.

**Comment by woody43567** [comment]
If you are hoping to use EntityFramework for your stores than currently you will have issues. Not entirely related to the number of instances but more related to the number activities. I stumbled across this issue when everything basically completely stopped working. When loading from the database it loads the activity instances with the workflow instance via a join which causes the scope field to multiply out by the number of activities. When you have a some variables/outputs  it can quickly blow out.  I got around this by implementing my own store for sql server (which is working but not production ready). I'm more of a Db kind of guy so I personally don't like using EF.  I don't believe  this is an issue if you are using a MongoDb store. There are also some other tricks you can do to help reduce the size of the scope field which can help.

Also, if you are using HttpActivities this may also cause you a problem with the Dashboard as currently checks for blocking activities on every request (currently what I am trying to work a way around). 

It is a great project which I will be using in a production environment, there are just a few optimization improvements I'm working on as I go through my proof of concept.

Please also keep in mind that it is also difficult to answer your question as there are many other factors to deal with when scaling.

Sorry for the long winded response but I hope this may help someone.

**Comment by generatives** [comment]
That is useful, thank you. My question was very broad but I think you and @sfmskywalker's answers provide a good view of how the system is doing in terms of scaling.

**Comment by steve99** [comment]
Currently looking at the Cosmos library to ensure it can scale too.  Discussing partition keys at the moment which are key to scalability.

**Comment by woody43567** [comment]
@steve99  I would be interested to hear your results with partitioning the data and what loads you have tested with.
Personally, I am looking at creating a workflow that automatically archives competed jobs to reduce the amount of data in the system. This works in my case (more long/large running workflows) as I don't really care about looking back at old successful jobs once it has passed a particular time frame.

**Comment by steve99** [comment]
@woody43567  Sure, still working on it, but hopeful.  I am looking at archiving also, but slightly more passive than yours, by creating partitions by time.  Will keep in touch.  The implementation wouldn't stop you moving to what you wanted.

**Comment by steve99** [comment]
Hi @woody43567 I have updated the code of Cosmos Persistence, so it deletes and recreates documents which have a status Partition key.  So far so good, but failed miserably in load testing as RU usage was exceeded when I fired queries as fast as my VM could issue them.  I will think about throttling, or a design which means timers are used from Service Bus or a Queue ensuring that all requests are taken, but Cosmos is not overwhelmed for the service you pay for.  I do similar in my core platform, where every x minutes it picks up n messages so it never gets overwhelmed, but devops have to intercept if the queue is too long, rarely as the VM's have peaks and troughs.
